### DISCUSSION TITLE
Why to add model & extraLibs to both JS & TS workers?

### DISCUSSION THREAD

**Discussion by manikantag** [microsoft/monaco-editor#3724] [Q&A]
Hi, 

When creating model or adding extraLib, why should we need to add to both JavaScript & TypeScript workers? Isn't JS worker just TS worker internally? On top of my head, this is increasing the memory footprint, as the current API creates model & extraLibs from the source string. 

_This may be answered already, but I couldn't find any similar questions or references online._ 

Thanks

**Comment by hediet** [comment]
> why should we need to add to both JavaScript & TypeScript workers?

What do you mean with need? Does the code currently do that or does your example not work if you don't do it?

**Reply by manikantag** [reply]
Please see  this [Playground link](https://microsoft.github.io/monaco-editor/playground.html?source=v0.36.1#XQAAAALyCAAAAAAAAABBqQkHQ5Njdzm2Nl5xIP9x6C0Suvw6CYToNDlyCv3AWEh1ldDst4RUTVoeqz8GJk1XmagPiVvXtreJAnV8PBLiDS2seUb0_VmCxYuEmp6d4u3eREE4_pQzuUKeHvrPD9VeUe9I5DlrKn2SIHwmGFuHVYoxIs23ctOhGAuz8ViFrga8F9YUok-K0V6lqxC28PaHyD9Qx6Xzi1K6PhGuhsv1fmDWqTCd3SoHvmjAa7Re5lm5f3n3WtQNpb4irKKPHOT2J0y7Op6F8bW8K5Gv1GethOFNeGgoRczYaSr-9C4jVNZGeP3vUah4dgXtR8-nIzRB4vbze-mlXsC8YDaD6wUVH0FrVzo31XiLWztZawAdCCbZaV3PmSRL3j0AW2G2GcNzsvLn3NVhKdqNQugTw1OdMmeLEvzta7vy8Q5c7cJPPrmQXn15QWdH3YXGN0BW_Gdvv8NDUY2j5rZtXRW8g8bKwbkYTiex8NaoFgTnOUkSUXBLJb4I5qKGB7xaN5Q6ATHNvlewhudyPzxpD6N7LlVsAgwo4bM3NFN3clQco8qZ7xKWUGAEX7Ynefuv0gTBFvhMkZYYp9PJvvyzF_VQR8lvHaoO47n2xBsn7WhFvWW4RnaQ2yBG66aqJoXdmSiLpKxXoVhuI3rTRNh2DLziy4YfhvHyt5DvkjqYDn5yUm1YkTTKQAP6lTSTyWZq4OK6KpgSfKKygfp-qoKFmbtDYIY_JVxwyHAeNMkVkhpzLa4F4MPBxyfL45D9uBgpLwRBwF_NYDwGWeyPTIAiFvAXKxEZh1_uS8PHAO-5BNM2qcoW2qbZmaufXoopNk73GIvmNQUIhBLDo6DjUnmGiIMdKKqGhDRDao6ncz_YnS2Sw-fPrX-cRFwNHJuGt0baxNxn5EmUHuVS7Ig_T8nANy-YmR_0rGYeg0jTgNCjNthd0WEFOHaWVkwN0my81nAN8u3yyvOjFjTilPmzmFBXvPle1wYq4qAdVHx5TdMKeNpeE0OYbdH9InJBxHJB80SXpjtbfj1iAJTZxmNyPzVlLlvRbDRm_XsEN7h0rUQUqcwp7dwto0piBdOJyqyf2g6VYO53jAAB2uQcAVnTtVsC4u_S_kWdRnRTzqMtLE6DQHVOPhnveHhfhtzDEVClBeOjaz3aFlXqTMoGP61ew_dpOKTuxlTXx3V368PDd7VW1A5nZvIDq5DK3UWS4H3_AsXWymtVLP6tE_oo23hXo06NkfiPLvq6DtGtfxicsQMrzMV-hP7ew8D7TwnzIzjQ8Fa7ST1NDC2fAhUYMndb_k8lvegTGrBsjfTr718FCJ53RSbTfd6nKU5jEza9KPjKVkbQJ3Z3_vlRuUY)

I created an extraLib _(Facts)_, a model _(TsAnimal)_ for TS worker and one model _(JsPerson)_ for JS worker. When you change the language, notice how the class defined in other language model doesn't recognized. 

~~extraLib, which is added to TS worker, works fine in either language.~~

**Reply by hediet** [reply]
That's a good question.

I don't know the typescript language feature implementation, but it looks like this is just not supported.
You should be able though to easily clone the monaco editor and debug this!
We'd be open for small pull requests (unless they cause breaking changes).

**Reply by manikantag** [reply]
Thanks for confirmation. Sure, I'll do my best.

**Reply by hediet** [reply]
Please look in Contributing.md and let me know about your debugging experience. I recently made improvements there with the hope that it will be easier for the community to investigate the bugs/feature requests they find.

**Reply by manikantag** [reply]
I stand corrected: extraLib also doesn't work always. Still needs to add to respective worker. 

In the above playground link, I was also creating the model from the definitions.

Fixed [playground link](https://microsoft.github.io/monaco-editor/playground.html?source=v0.36.1#XQAAAALGCQAAAAAAAABBqQkHQ5Njdzm2Nl5xIP9x6C0Suvw6CYToNDlyCv3AWEh1ldDst4RUTVoeqz8GJk1XmagPiVvXtreJAnV8PBLiDS2seUb0_VmCxYuEmp6d4u3eREE4_pQzuUKeHvrPD9VeUe9I5DlrKn2SIHwmGFuHVYoxIs23ctOhGAuz8ViFrga8F9YUok-K0V6lqxC28PaHyD9Qx6Xzi1K6PhGuhsv1fmDWqTCd3SoHvmjAa7Re5lm5f3n3WtQNpb4irKKPHOT2J0y7Op6F8bW8K5Gv1GethOFNeGgoRczYaTA_zlDvOdBN-Ndrh1rqiF_8Zci3-UdWamgyHmG_FxrV9VXhsrarvRFhUxHMGAP7RNDZoLMpqxH8v2e1iF-JdrATISSZGUAhwXUm4uGsjLkZD-GaslCEHmCNMSX9fP6oBP7f-Ko0MSc0HMkNRWYLiqSmjgiu3-MILe9cba86Qna1rSlSg-xCu7CNIDr6jGE8T3mAk8VlJ0actAsJ00khAIb69ma8dAO65C0_Oo3fnEy6jD-Z7gzHu0K2JHpkOlVl5Gzn2gIP8m0FOsVDtIRya8xfr-HN21birn4vwOyB-9PCad2DXEXuMRqaeoiFaECrR-aYvcQD_JHT8XUsMQgH6nJEwGDJrkTAsOpEODJg_GUS45AvmKB1N57oKfKuZ6YCxWC1Q250UaSEPw_O_BqeYhJ9xc3JoDoh3s_c6zgtbcF6Cv4BkP00kUThAjE7QKuBpL64qAsL41nEqg2eeacPeXUpWvtDHMrHJiuArAVa8Dgw0ryjxS5uuFixIlu4y2wydgDaTj6jmFk4hJz0VBS5Vj8AZtSoPL30zYHV3Hwb54UT9D4TMgS_TgkmrSRDJjovuOLJAbOTPeGhQLz3jC4XHiYjpjvd0csXknp91bEWIhbsM4-iPUNJnMkmZTH4nPsPGRYsuESQZM6lNJZU7JqTzOV2w7WBClvb41W14CpyQatLX96tyb4N13kROvBeeEtr6KZsp-a7TaBfDY_eWo9aGXdldml5A3ABDs1cG3vxqsdhp_pA9qAuUnL9ShUiWVGKVTsogN-qVb4b1eOpAcbktkwrJa1-SieEvh-cXY5rW6kEa4CFfCMvkHDQHVeLV7G-GaeO-rfvOsQMZXQfzYJ9cPHT0OZ71oiWev7u_dQiBLolw8GYctKJFUNXTr8iWRXySpZEHAdS0Fn0GboV4_DVHob_R8QRU11ZamQk7lpLKF78-E_m8H8BLUhGU0lnwQpOlYmPVC_ZVfsGEbtJKXAQmAfHv3OrmeaRM4vAmhcURhxqAf_WrBIE)

**Comment by manikantag** [comment]
I'll document my findings & progress here:

Some discussion about separate workers: https://github.com/Microsoft/monaco-typescript/pull/8#discussion_r73651581
